### PR TITLE
GH203 Split groupby with as_index (temptative)

### LINKED ISSUE #203: BUG: df.groupby(col, as_index=False).value_counts() returns a DataFrame but is annotated as Series
**Describe the bug**

Using 
```
df = pd.DataFrame({"a": [1, 2, 3], "b": 1})

x: pd.DataFrame = df.groupby("a", as_index=False).value_counts()
```
raises an error, because mypy thinks the return is a Series

```
 Incompatible types in assignment (expression has type "Series[int]", variable has type "DataFrame")  [assignment]
``` 

**To Reproduce**
1. Provide a minimal runnable `pandas` example that is not properly checked by the stubs.
2. Indicate which type checker you are using (`mypy` or  `pyright`).
3. Show the error message received from that type checker while checking your example.

```
df = pd.DataFrame({"a": [1, 2, 3], "b": 1})

x: pd.DataFrame = df.groupby("a", as_index=False).value_counts()
```

In theory this should be ok, but not sure if this ``as_index=False`` can be typed?

**Please complete the following information:**
 - OS:  ubuntu
 - OS Version 22.04
 - python version 3.10
 - version of type checker mypy 0.971
 - version of installed `pandas-stubs` newest 20220815

**Additional context**

### PR DESCRIPTION
**Pull request by loicdiridollou**
- [x] Closes #203 
- [x] Tests added: Please use `assert_type()` to assert the type of any return value

### CODE REVIEW

**Comment by loicdiridollou** [comment]
I am still having overlap issues in the overload but looking into it, feel free to suggest something else.
Right now just done for the Scalar but can/should be expanded to the other types.

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
Good start, but I think you will have to work on the following:

- Have to follow the suggestion at https://github.com/pandas-dev/pandas-stubs/issues/203#issuecomment-1218439288 , item 2, to add a `Generic` parameter to `DataFrameGroupBy` that essentially represents whether `as_index` is `True` or `False`
- You will have to do the same separation on the overloads for `groupby()` with the other `by` arguments, and add appropriate tests.

**Comment by loicdiridollou** [comment]
Seems like this is going to take a little longer, I think I understand your suggestion in #203, let me try to come up with something.

File: tests/test_frame.py [hunk 1]
```diff
@@ -1048,7 +1066,7 @@ def test_types_groupby() -> None:
 
     df1: pd.DataFrame = df.groupby(by="col1").agg("sum")
     df2: pd.DataFrame = df.groupby(level="ind").aggregate("sum")
-    df3: pd.DataFrame = df.groupby(by="col1", sort=False, as_index=True).transform(
+    df3: pd.Series = df.groupby(by="col1", sort=False, as_index=True).transform(
```
[the remark this answers is not in the document]
Comment (loicdiridollou) [reply tests/test_frame.py]: Agreed, I will tackle those once we agree on the potential solution so I can do one thing at a time not to confuse the amount of code to review too much.

File: pandas-stubs/core/frame.pyi [hunk 2]
```diff
@@ -229,6 +235,11 @@ class _LocIndexerFrame(_LocIndexer):
         value: Scalar | NAType | NaTType | ArrayLike | Series | list | None,
     ) -> None: ...
 
+_TT = TypeVar("TT", bound=Literal[True, False])
+
+class DataFrameGroupByGen(DataFrameGroupBy[ByT], Generic[ByT, _TT]): ...
```
Comment (loicdiridollou) [inline pandas-stubs/core/frame.pyi:240]: @Dr-Irv these seem to be more what you had originally in mind, please confirm if that is what you were thinking

File: pandas-stubs/core/groupby/groupby.pyi [hunk 3]
```diff
@@ -235,8 +239,10 @@ class GroupBy(BaseGroupBy[NDFrameT]):
     @final
     @overload
     def size(self: GroupBy[Series]) -> Series[int]: ...
-    @overload  # return type depends on `as_index` for dataframe groupby
-    def size(self: GroupBy[DataFrame]) -> DataFrame | Series[int]: ...
+    @overload
+    def size(self: DataFrameGroupByGen[ByT, Literal[True]]) -> Series[int]: ...  # type: ignore[misc]
```
Comment (loicdiridollou) [inline pandas-stubs/core/groupby/groupby.pyi:243]: Surprisingly here `mypy` throws the error that the DataFrameGroupByGen is not a supertype of DataFrameGroupByGen, I have ignored the error but open to a better solution.

File: pandas-stubs/core/frame.pyi (hunk 2 above)
Comment (Dr-Irv) [reply pandas-stubs/core/frame.pyi:240]: No, I was thinking of doing this:
```python
class DataFrameGroupBy(GroupBy[DataFrame], Generic[ByT, _TT]):
```
No need to introduce a new class

File: pandas-stubs/core/frame.pyi [hunk 4]
```diff
@@ -229,6 +235,11 @@ class _LocIndexerFrame(_LocIndexer):
...
+class SeriesGroupByGen(SeriesGroupBy, Generic[_TT, ByT]): ...
```
Comment (Dr-Irv) [inline pandas-stubs/core/frame.pyi:241]: Same here - you can have more than one parameter to the generic nature.

File: pandas-stubs/core/groupby/groupby.pyi (hunk 3 above)
Comment (Dr-Irv) [reply pandas-stubs/core/groupby/groupby.pyi:243]: I think the solution here is to push `def size` down to the `DataFrameGroupBy` and `SeriesGroupBy` classes, and remove it from the `GroupBy` class.

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
Hope my comments help.  Note - I will be not be able to give feedback until after Wed., October 23.

**Review by Dr-Irv — CHANGES_REQUESTED** [review]
I think this is looking good, but CI is failing due to overlapping overload.  You'll have to use `# type: ignore` statements to get around that.

**Comment by Dr-Irv** [comment]
So I am fine with this.  Do you want me to merge it so that `size()` is fixed?  Or do you want to work on other methods where the type of the result depends on `as_index` parameter?

**Comment by loicdiridollou** [comment]
I will open another issue to track down the other methods, would rather merge small changes than one large PR.

**Review by Dr-Irv — APPROVED** [review]
thanks @loicdiridollou

**Pull request merged**